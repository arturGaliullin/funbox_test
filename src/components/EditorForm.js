import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class EditorForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            placemark: '',
            placemarkData: null
        }
    }

    onHandleChange = (e) => {
        this.setState({
            placemark: e.target.value
        }, async () => {
            await this.props.searchData(this.state.placemark);
        });
    };

    onSelect = (name, data) => {
        this.setState({
            placemark: name,
            placemarkData: data
        }, () => {
            this.props.clearData();
        });
    };

    onSave = (e) => {
        e.preventDefault();
        this.setState({
            placemark: ''
        }, () => {
            if (this.state.placemarkData !== null && this.state.placemarkData.name) {
                this.props.addPlacemark(this.state.placemarkData);
            } else if (this.props.results.length > 0) {
                this.props.addPlacemark(this.props.results[0]);
                this.props.clearData();
            }
        });

    };

    render() {
        const { results } = this.props;

        return (
            <form onSubmit={this.onSave} className='editor__form'>
                <input
                    className='editor__form-input'
                    type="text"
                    onChange={this.onHandleChange}
                    value={this.state.placemark}
                    required={true}
                    name='placemark'
                />
                <div className='editor__form-results'>
                    {results.length > 0 ? results.map(result =>
                        <div
                            key={result.id}
                            className='editor__form-result'
                            onClick={() => this.onSelect(result.name, result)}
                        >
                            {result.name}
                        </div>
                    ) : null}
                </div>
                <button onClick={this.onSave} className="editor__form-button">Добавить</button>
            </form>
        );
    }
}

EditorForm.propTypes = {
    clearData: PropTypes.func,
    searchData: PropTypes.func,
    addPlacemark: PropTypes.func,
    results: PropTypes.array
};
