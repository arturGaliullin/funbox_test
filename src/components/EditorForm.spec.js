import React from 'react';
import { shallow, mount } from 'enzyme';
import EditorForm from './EditorForm';

const props = {
    clearData: () => {},
    searchData: () => {},
    addPlacemark: () => {},
    results: []
};

const state = {
    placemark: '',
    placemarkData: null
}

describe('EditorForm', () => {
    it('should render with props', () => {
        const component = shallow(<EditorForm {...props} {...state} />);

        expect(component).toMatchSnapshot();
    });

    it('check onChange', () => {
        const component = shallow(<EditorForm {...props} {...state} />);

        component.find('.editor__form-input').simulate('change', {target: {value: 'value'}});
        expect(component.state('placemark')).toBe('value');
    });

    it('check result onClick', () => {
        props.results = [{id: 1, name: 'name'}, {id: 2, name: 'name-1'}];
        let componentMount = mount(<EditorForm {...props} {...state} />);

        componentMount.find('.editor__form-result').first().simulate('click');
        componentMount.find('.editor__form-input').instance().value = props.results[0].name;

        state.placemarkData = props.results[0];
        state.placemark = props.results[0].name;

        expect(componentMount.find('.editor__form-input').props().value).toBe(props.results[0].name);
        expect(componentMount.state('placemarkData')).toBe(props.results[0]);
        expect(componentMount.state('placemark')).toBe(props.results[0].name);

        props.results = [];

        componentMount = mount(<EditorForm {...props} {...state} />);
        expect(componentMount.props().results).toHaveLength(0);
    });

});