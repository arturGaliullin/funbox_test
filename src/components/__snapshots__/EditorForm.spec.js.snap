// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorForm should render with props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditorForm
    addPlacemark={[Function]}
    clearData={[Function]}
    placemark=""
    placemarkData={null}
    results={Array []}
    searchData={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="editor__form-input"
          name="placemark"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />,
        <div
          className="editor__form-results"
        />,
        <button
          className="editor__form-button"
          onClick={[Function]}
        >
          Добавить
        </button>,
      ],
      "className": "editor__form",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "editor__form-input",
          "name": "placemark",
          "onChange": [Function],
          "required": true,
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "className": "editor__form-results",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Добавить",
          "className": "editor__form-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Добавить",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="editor__form-input"
            name="placemark"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />,
          <div
            className="editor__form-results"
          />,
          <button
            className="editor__form-button"
            onClick={[Function]}
          >
            Добавить
          </button>,
        ],
        "className": "editor__form",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "editor__form-input",
            "name": "placemark",
            "onChange": [Function],
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "editor__form-results",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Добавить",
            "className": "editor__form-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Добавить",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
